examples/trait-erc20/lib.rs

```diff
@@ -0,0 +1,495 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_env::{
+    DefaultEnvironment,
+    Environment,
+};
+use ink_lang as ink;
+
+pub use erc20::Erc20;
+
+type Balance = <DefaultEnvironment as Environment>::Balance;
+type AccountId = <DefaultEnvironment as Environment>::AccountId;
+
+/// The ERC-20 error types.
+#[derive(Debug, PartialEq, Eq, scale::Encode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum Error {
+    /// Returned if not enough balance to fulfill a request is available.
+    InsufficientBalance,
+}
+
+/// The ERC-20 result type.
+pub type Result<T> = core::result::Result<T, Error>;
+
+/// Trait implemented by all ERC-20 respecting smart contracts.
+#[ink::trait_definition]
+pub trait BaseErc20 {
+    #[ink(constructor)]
+    fn new(initial_supply: Balance) -> Self;
+
+    #[ink(message)]
+    fn total_supply(&self) -> Balance;
+
+    #[ink(message)]
+    fn balance_of(&self, owner: AccountId) -> Balance;
+
+    #[ink(message)]
+    fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance;
+
+    #[ink(message)]
+    fn transfer(&mut self, to: AccountId, value: Balance) -> Result<()>;
+
+    #[ink(message)]
+    fn approve(&mut self, spender: AccountId, value: Balance) -> Result<()>;
+
+    #[ink(message)]
+    fn transfer_from(
+        &mut self,
+        from: AccountId,
+        to: AccountId,
+        value: Balance,
+    ) -> Result<()>;
+}
```

Comment:
We could also move this entire section into the contract module so that we do not have to specify `Balance` and `AccountId` which is ugly. Also we wouldn't need to import of the trait. What do you think? Which is better as learning material?

Reply:
I think we should omit anything that is not strictly necessary to demonstrate basic Trait functionality. As to make it easily understandable and reduce causes for confusion. Yeah, so I'll adapt it.